kphp-inspector/src/file_searcher.js

```diff
@@ -38,42 +39,139 @@ function doesFileContainFunction(fullFileName) {
 
-  return output.filter(fn => doesFileSatisfyQ(fn, qParts) && doesFileContainFunction(fn));
+  return files.filter(fn => doesFileSatisfyQ(fn, qParts) && doesFileContainFunction(fn));
 }
 
 /**
- * @param {string} q Search string from console
+ * @param {string} query Search string from console
  * @return {string[]} Full file names, to be passed to parseCppClass()
  */
-function performSearchForClass(q) {
-  let qParts = splitUserSearchStr(q);
-  if (!qParts.length) {
-    return [];
+function performSearchForClass(query) {
+  const strict = query.startsWith('\\');
+  utils.debug(`Start search class by query: ${query}, strict: ${strict}`);
+
+  if (strict) {
+    // Convert the name to the form that KPHP uses for files.
+    const className = convertFqnToCppFileName(query);
+    const classnameWithDot = className + '\\.';
+    utils.debug(`Filename for search: ${classnameWithDot}`);
+
+    const files = findFilesWithNameLike(classnameWithDot);
+    utils.debug(`Found files: ${files.length === 0 ? 'nothing' : files.join(', ')}`);
+
+    return files.filter(file => file.endsWith('.h'));
+  }
+
+  let { qParts, longestQ } = queryToParts(query);
+  let files = findClassFilesWithNameLike(longestQ);
+  utils.debug(`Found files: ${files.length === 0 ? 'nothing' : files.join(', ')}`);
+
+  return files.filter(file => file.endsWith('.h') && doesFileSatisfyQ(file, qParts));
+}
+
+/**
+ * @param {string} query
+ * @returns {(string[]|string)[]}
+ */
+function queryToParts(query) {
+  let qParts = splitUserSearchStr(query);
+  if (qParts.length === 0) {
+    return [[], ''];
   }
+
+  utils.debug(`Query parts: ${qParts.join(', ')}`);
+
   let longestQLen = Math.max(...qParts.map(p => p.length));
   let longestQ = qParts.find(p => p.length === longestQLen);
+  utils.debug(`Longest query part: ${longestQ}`);
+  return [qParts, longestQ];
+}
+
+/**
+ * @param {string} query
+ * @return {string[]}
+ */
+function findClassFilesWithNameLike(query) {
+  return findFilesWithNameLike(query, '/cl');
+}
+
+/**
+ * @param {string} query
+ * @param {boolean} searchInNames
+ * @return {string[]}
+ */
+function findFunctionFilesWithNameLike(query, searchInNames = true) {
+  return findFilesWithNameLike(query, searchInNames).filter((filename, _, output) =>
+    // strip out duplicates and strange files: leave only .cpp and .h, also filter out .h files of classes
+    filename.endsWith('.cpp') ||
+    (
+      filename.endsWith('.h') &&
+      !output.includes(filename.substr(0, filename.length - 2) + '.cpp') &&
+      !filename.includes('cl/C@')
+    ),
+  );
+}
+
+/**
+ * @param {string} query
+ * @param {boolean} searchInNames
+ * @param {string} pathSuffix
+ * @return {string[]}
+ */
+function findFilesWithNameLike(query, searchInNames = true, pathSuffix = '') {
+  const command = `find ${env.RUN_ARGV.KPHP_COMPILED_ROOT}${pathSuffix} ${searchInNames ? "-iname" : "-iwholename"} "*${query}*"`;
```

Comment:
The name `searchInNames` doesn't describe what this parameter actually does

Reply:
Fixed.